app/routes.py

```diff
@@ -1,2 +1,245 @@
+    task.completed_at = datetime.utcnow()
+
+    return jsonify({ "task" : {
+        "id" : task.task_id,
+        "title" : task.title,
+        "description" : task.description,
+        "is_complete" : task.is_complete
+    }}), 200
+
+@tasks_bp.route("/<task_id>/mark_incomplete", methods=["PATCH"])
+def mark_tasks_incomplete(task_id):
+    task = Task.query.get(task_id)
+
+    if task is None:
+        return jsonify(None), 404
+
+    # request_body = request.get_json()
+
+    task.is_complete = False
+    task.completed_at = None
+
+    return jsonify({ "task" : {
+        "id" : task.task_id,
+        "title" : task.title,
+        "description" : task.description,
+        "is_complete" : task.is_complete
+    }}), 200
+
+@goals_bp.route("", methods=["GET","POST"])
+def handle_goals():
+    if request.method == "GET":
+        goals = Goal.query.all()
+
+        goals_response = []
+        for goal in goals:
+            goals_response.append({
+                "id" : goal.goal_id,
+                "title" : goal.title
+            })
+        return jsonify(goals_response)
+
+    elif request.method == "POST":
+        request_body = request.get_json()
+        title = request_body.get("title")
+
+        if not title:
+            return jsonify({
+                "details" : "Invalid data"
+            }), 400
+
+        new_goal = Goal(title=title)
+
+        db.session.add(new_goal)
+        db.session.commit()
+
+        return jsonify({"goal" : {
+            "id" : new_goal.goal_id,
+            "title" : new_goal.title
+        }}), 201
+
+@goals_bp.route("/<goal_id>", methods=["GET", "PUT", "DELETE"])
+def handle_goal(goal_id):
+    goal = Goal.query.get(goal_id)
+
+    if not goal:
+        return jsonify(None), 404
+    
+    if request.method == "GET":
+        return jsonify({"goal" : {
+            "id" : goal.goal_id,
+            "title" : goal.title
+        }})
+    
+    elif request.method == "PUT":
+        request_body = request.get_json()
+        title = request_body.get("title")
+
+        if title:
+            goal.title = title
+
+            db.session.commit()
+
+            return jsonify({"goal" : {
+                "id" : goal.goal_id,
+                "title" : goal.title
+            }})
+        else:
+            return jsonify(None), 404
+    
+    elif request.method == "DELETE":
+        db.session.delete(goal)
+        db.session.commit()
+
+        return jsonify({
+            "details" : f'Goal {goal.goal_id} "{goal.title}" successfully deleted'
+        }) 
+
+@goals_bp.route("/<goal_id>/tasks", methods=["POST", "GET"])
+def handle_goals_tasks(goal_id):
+    goal = Goal.query.get(goal_id)
+    if goal is None:
+        return jsonify(None), 404
+
+    if request.method == "POST":
+        request_body = request.get_json()
+
+        task_ids = request_body.get("task_ids")
+        for task_id in task_ids:
+            task = Task.query.get(task_id)
+            task.goal = goal
```

Comment:
👍 

This is absolutely a valid way to associate a task with a goal. But we can also think about going from the opposite side of the relationship. We could write
```python
            goal.tasks.append(task)  # append the looked up task to the tasks collection of the goal.
```

In either situation, we should also think about whether to interpret POSTing to this endpoint as setting the exact list of tasks that should be a part of the goal, or whether we are adding these tasks to the goal (in addition to anything that was already there). Neither the project description nor the test `test_post_task_ids_to_goal_already_with_goals` are entirely clear on the desired behavior. Which perspective does your implementation take? How could we implement the other approach?